R/rowClust.R

```diff
@@ -0,0 +1,83 @@
+#' Clusters rowData using \code{fastcluster::hclust}
+#'
+#' \code{rowClust} Performs hierarchical clustering on rowData of the 
+#' \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}, 
+#' by using the dissimilarity function provided at \code{dissimilarity_FUN}.
+#'
+#' @param x a
+#'   \code{\link[SummarizedExperiment:SummarizedExperiment-class]{SummarizedExperiment}}
+#'   object.
+#'   
+#' @param abund_values a \code{character} vector, name of the assay containing the data.
+#' 
+#' @param name a \code{character} vector, name of the new column at \code{rowData} where to place cluster ids.
+#' 
+#' @param clustering_method a \code{character} vector representing the clustering 
+#' method to be used, same as methods available at \code{\link[hclust]{hclust}}.
+#' 
+#' @param transposed \code{logical} whether the data should be transposed.
+#' 
+#' @param dissimilarity_FUN \code{stats::cor} function to compute the dissimilarity matrix
+#' 
+#' @param ... other arguments passed onto \code{\link[cutree]{cutree}}
+#'
+#' @return \code{x} with additional \code{\link{rowData}} named \code{*name*}
+#'
+#' @export
+#'
+#' @examples
+#' # Load example data
+#' library(mia)
+#' data("GlobalPatterns")
+#' tse <- GlobalPatterns
+#' # Agglomerate to Class rank; for speeding up the example.
+#' tse <- agglomerateByRank(tse, rank="Class")
+#' # Cluster using Spearman correlation as a dissimilarity and ward.D2 
+#' # for clustering; with results having 6 clusters
+#' tse <- rowClust(tse, clustering_method = "ward.D2",
+#'                 dissimilarity_FUN = function(x) stats::cor(x, method="spearman"),
+#'                 k = 6)
+#' # Aggregate clusters as a sum of each cluster values
+#' # (CAG: Co-Abundance Groups)
+#' tse_CAG <- mergeRows(tse, rowData(tse)$clust_id)
+#'
+NULL
+
+#' @importFrom fastcluster hclust
```

Comment:
accoding to automated checks, this seems to conflict `importFrom stats hclust`. That is used in `getExperimentCrossCorrelation` by @TuomasBorman - I think we could use fastcluster implementation there as well. If Tuomas agrees, this can be changed as a *separate* PR.

Reply:
Hmmm... does this `fastcluster` give us any additional value?  `stats` is default package, so by using `stats::hclust` we do not create any extra dependencies

But if that `factcluster` is better and faster, then I'm fine with that, and we can use it also in `getExperimentCrossCorrelation`

Reply:
It is better and faster [according to their own benchmarks](https://www.jstatsoft.org/article/view/v053i09).

At least miaverse will get better in terms of CO2 emission contributions if we optimize these things..

If these two considerations are crucial enough to justify more dependencies is another question. The same solution should be used in all functions, no matter which option we end up choosing.

I see 2-5 fold speedups with some of our own demo data sets. See e.g. below. For most practical use cases this probably does not make huge difference because the computing is not prohibitively slow in either case. Hence, I have no strong opinions but perhaps a slight preference to the computationally more efficient solutions if this can be easily added.


```
library(microbiomeDataSets)
tse <- GrieneisenTSData()
x <- assay(tse, "counts")
d <- dist(x)
ts <- system.time(hcs <- stats::hclust(d))
tf <- system.time(hcf <- fastcluster::hclust(d))
print(ts)
print(tf)


```

Reply:
Drops make an ocean, I guess...

I prefer to keep this as simple as possible and to avoid dependencies that are not vital, but speed is also thing that we must consider. 

In `getExperimentCrossCorrelation`, other things create the bottle neck, and this clustering part is not something that requires this kind of optimization. But here I think it is more beneficial to have faster algorithm...

However, this is very easy to change: just `stats::hclust()` --> `fastcluster::hclust()`

--> It's ok with me if we start using `fastcluster`

Reply:
Hmm also true that one can give fastcluster::hclust as a function argument.

Wouldn't that be a solution: use hclust by default to avoid dependencies, and show in examples how to speed up with fastcluster?

Reply:
I like that idea 

- There could be an internal function that has argument `clustering_method = stats::hclust`
     - By using that kind of approach, we could hide argument with `...` so that there are not too many options
     - Internal function should have input check that ensures that method is `stats::hclust` or `fastcluster::hclust`, because we cannot be sure that all methods return results in same format

Reply:
This is ok, let's do.

The clustering should use CLR-transformed data in order to reduce compositionality bias. This brings it closer to SparCC and SPIEC-EASI, which are taxonomic co-occurrence network reconstuction methods. We may need to compare to these later, but clr + hclust should be good enough for now, and I have seen it occasionally being used in literature.

Reply:
Should this kind of transformation (clr) be optional? Would the method also be meant for other type of data? (e.g functional, metabolomic)

Reply:
Yes, of course. True. But make sure to have this in the examples.

Reply:
> I like that idea
> 
>     * There could be an internal function that has argument `clustering_method = stats::hclust`
>       
>       * By using that kind of approach, we could hide argument with `...` so that there are not too many options
>       * Internal function should have input check that ensures that method is `stats::hclust` or `fastcluster::hclust`, because we cannot be sure that all methods return results in same format

Regarding the implementation of it. I have tried to compare the functions (provided and the options allowed), then the comparison/checking got lengthy and messy. However, at the end of it, since both  `stats::hclust` and `fastcluster::hclust` return an object of class `hclust`, I used that as the checking criterion for accepting the clustering function provided.

Reply:
Is it clear now? I think we just use stats::hclust as the default, but show an example with fastcluster::hclust. OK?

Reply:
Sure, I'll add a fastcluster::hclust example.